lib/constraint_solver/generalization.ml

```diff
@@ -27,59 +37,132 @@ module Region = struct
   end
 end
 
+module Guard = Identifier
```

Comment:
The guard representation should be improved in a future PR. 

Namely there are two kinds of guards we mainly employ -- direct guards (matches to types) and indirect guards (partial generics to instances). The first has a stable identifier whereas the latter doesn't (the identifier of a partial generic evolves due to unification)